database/pgx/pgx.go

```diff
@@ -298,9 +298,21 @@ func (p *Postgres) runStatement(statement []byte) error {
 			if pgErr.Detail != "" {
 				message = fmt.Sprintf("%s, %s", message, pgErr.Detail)
 			}
-			return database.Error{OrigErr: err, Err: message, Query: statement, Line: line}
+			migrationErr = database.Error{OrigErr: err, Err: message, Query: statement, Line: line}
+		} else {
+			migrationErr = database.Error{OrigErr: err, Err: "migration failed", Query: statement}
+		}
+
+		// For safety, always issue a rollback on error. In multi-statement
+		// mode, this is necessary to make sure that the connection is not left
+		// in an aborted state. In single-statement mode, this will be a no-op
+		// outside of the implicit transaction block that was already rolled
+		// back.
+		if _, rollbackErr := p.conn.ExecContext(ctx, "ROLLBACK"); rollbackErr != nil {
+			rollbackErr = fmt.Errorf("failed to rollback migration tx: %w", rollbackErr)
+			return multierror.Append(migrationErr, rollbackErr)
```

Comment:
I can do enough CFA-by-eye to know that `migrationErr` can't be `nil` here, but I wonder if we could protect against that my moving `migrationErr = database.Error{OrigErr: err, Err: "migration failed", Query: statement}` to the initialization of `migrationErr` - or if we'd even want to?

Reply:
One thing of note is that hashicorp (in their infinite wisdom) has made `Append` work such that if `migrationErr` _were_ `nil` here it would still at least generate a `multierror` with a single error inside of it: https://github.com/hashicorp/go-multierror/blob/v1.1.1/append.go#L36-L39

à la https://go.dev/play/p/HOmHw7-3Dbf

That being said, defining `migrationErr` from the get-go as the version derived from a non-`pgconn` error to start with and overwriting with a more specific `database.Error` when possible is fine by me as well! 

Switched over accordingly in https://github.com/nicheinc/migrate/pull/1/commits/4813cdb3d9e407aef2aafc9fd1fcf3d12a6e8f71

Reply:
One thing of note is that hashicorp (in their infinite wisdom) has made `Append` work such that if `migrationErr` _were_ `nil` here it would still at least generate a `multierror` with a single error inside of it: https://github.com/hashicorp/go-multierror/blob/v1.1.1/append.go#L36-L39

à la https://go.dev/play/p/HOmHw7-3Dbf

That being said, defining `migrationErr` from the get-go as the version derived from a non-`pgconn` error to start with and overwriting with a more specific `database.Error` when possible is fine by me as well! 

Switched over accordingly in https://github.com/nicheinc/migrate/pull/1/commits/4813cdb3d9e407aef2aafc9fd1fcf3d12a6e8f71